File: Problem2_KClosestElements.java
// Time Complexity : o(n)
// Space Complexity : o(1)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No

class Solution {
    public List<Integer> findClosestElements(int[] arr, int k, int x) {
        
        List<Integer> result=new ArrayList<>();
        
        if(arr==null || arr.length==0)
            return result;
        
        int low=0;
        int high=arr.length-1;
        
        
        while(high-low+1>k)
        {
            int distLeft=x-arr[low];
            int distRight=arr[high]-x;
            
            if(distLeft>distRight)
            {
                low++;
            }
            if(distLeft<=distRight)
            {
                high--;
            }
        }
        
        for(int i=low;i<=high;i++)
            result.add(arr[i]);
        
        return result;
    }
}

File: Problem-2-k-closest elements.py
class Solution:

    def findClosestElements(self, arr, k, x) :

        result = []
        #### two pinters approach
        if arr == None or len(arr) == 0:
            return result
        low = 0
        high = len(arr) - 1
        while (high - low + 1) > k:
            disL = x - arr[low]
            disR = arr[high] - x

            if disL > disR:
                low += 1
            else:
                high -= 1
        for i in range(low, high + 1):
            result.append(arr[i])

        return result

        #### range binary search
        # base
        if arr == None or len(arr) == 0:
            return result

        # logic
        low = 0
        high = len(arr) - k

        while low < high:
            mid = (low + high) // 2

            disL = x - arr[mid]
            disR = arr[mid + k] - x

            if disL > disR:
                low = mid + 1
            else:
                high = mid

        for i in range(low, low + k):
            result.append(arr[i])
        return result